source/common/http/http1/codec_impl.cc

```diff
@@ -785,7 +793,13 @@ Envoy::StatusOr<int> ConnectionImpl::onHeadersCompleteBase() {
 }
 
 void ConnectionImpl::bufferBody(const char* data, size_t length) {
-  buffered_body_.add(data, length);
+  auto slice = current_dispatching_buffer_->frontSlice();
+  if (data == slice.mem_ && length == slice.len_) {
+    buffered_body_.move(*current_dispatching_buffer_, length);
+    dispatching_slice_already_drained_ = true;
+  } else {
+    buffered_body_.add(data, length);
+  }
```

Comment:
One possible concern is that the new code won't reduce the size of slices in the buffer in cases where the read from the socket produced less than 4032 bytes.  After this change, the slice in buffered_body_ would be roughly 20kb instead of 4kb.  But I don't think that would make trickle memory attacks significantly more effective.  If we were really concerned we could limit this optimization to the case where buffered_body_ was empty when we entered innerDispatch.

Reply:
My feeling is that it's probably ok because we had already allocated the over-sized 20k slice. We're now just not right-sizing it here, but not making the memory use higher. So it's a minor regression on memory use for this case, but actually results in a slightly lower peak use (because we don't need the 4k slice to be allocated before free'ing the 20k slice). 

But if anyone feels differently, we can look at heuristically doing a copy here.

Reply:
I agree with you.

Reply:
The `buffered_body_` is drained after all slices have been processed, so the overhead in that buffer should not matter.

Reply:
Fragmentation in buffered_body_ is transferred to the output buffer since slices are moved from one buffer to the other.  Still, I think the attack potential here is relatively low.